src/main/java/com/gotocompany/depot/maxcompute/converter/payload/MessagePayloadConverter.java

```diff
@@ -0,0 +1,44 @@
+package com.gotocompany.depot.maxcompute.converter.payload;
+
+import com.aliyun.odps.data.SimpleStruct;
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.Message;
+import com.gotocompany.depot.maxcompute.converter.type.MessageTypeInfoConverter;
+import lombok.RequiredArgsConstructor;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+@RequiredArgsConstructor
+public class MessagePayloadConverter implements PayloadConverter {
+
+    private final MessageTypeInfoConverter messageTypeInfoConverter;
+    private final List<PayloadConverter> payloadConverters;
+
+    @Override
+    public Object convertSingular(Descriptors.FieldDescriptor fieldDescriptor, Object object) {
+        Message dynamicMessage = (Message) object;
+        List<Object> values = new ArrayList<>();
+        Map<Descriptors.FieldDescriptor, Object> payloadFields = dynamicMessage.getAllFields();
+        fieldDescriptor.getMessageType().getFields().forEach(innerFieldDescriptor -> {
+            if (!payloadFields.containsKey(innerFieldDescriptor)) {
+                values.add(null);
+                return;
+            }
+            Object mappedInnerValue = payloadConverters.stream()
+                    .filter(converter -> converter.canConvert(innerFieldDescriptor))
+                    .findFirst()
+                    .map(converter -> converter.convert(innerFieldDescriptor, payloadFields.get(innerFieldDescriptor)))
+                    .orElse(null);
+            values.add(mappedInnerValue);
+        });
+        return new SimpleStruct(messageTypeInfoConverter.convertSingular(fieldDescriptor), values);
+    }
```

Comment:
The converter creates new ArrayList for each conversion which is inefficient. Also, the stream operations inside a loop can cause performance issues. Consider using object pooling and optimizing the converter lookup.

Reply:
could you please tell me which particular object to pool?